File: packages/node/src/browser.ts
import { dirname, resolve } from "node:path";

import {
  requireFrameworkHumanizeSupport,
  resolveFrameworkVersion,
  type FrameworkBackend,
} from "./automation.js";
import { ConfigurationError } from "./errors.js";
import { prepareLaunch } from "./launcher.js";
import { resolveHumanConfig, type HumanConfig, type HumanPreset } from "./humanize.js";

export interface ChromiumLauncherLike<TBrowser = unknown, TContext = unknown> {
  launch(options: Record<string, unknown>): Promise<TBrowser>;
  launchPersistentContext(userDataDir: string, options: Record<string, unknown>): Promise<TContext>;
}

export interface PlaywrightLike<TBrowser = unknown, TContext = unknown> {
  chromium: ChromiumLauncherLike<TBrowser, TContext>;
}

export interface PuppeteerLike<TBrowser = unknown> {
  launch(options: Record<string, unknown>): Promise<TBrowser>;
}

export interface FrameworkLaunchSettings {
  profile?: Record<string, unknown>;
  launchOptions?: Record<string, unknown>;
  tempRoot?: string;
  nativeReady?: boolean;
  nativeReadyTimeout?: number;
  humanize?: boolean;
  humanPreset?: HumanPreset;
  humanConfig?: Partial<HumanConfig>;
  humanSeed?: number;
  frameworkVersion?: string;
  runtimeHandoff?: Record<string, unknown>;
  allowRuntimeActivationTicket?: boolean;
  releaseRoot?: string;
}

type JsonObject = Record<string, unknown>;

function jsonObject(value: unknown): JsonObject | undefined {
  return value && typeof value === "object" && !Array.isArray(value) ? value as JsonObject : undefined;
}

function parseSignedLeaseClaims(lease: string | Buffer | Record<string, unknown>): JsonObject | undefined {
  try {
    const envelope = typeof lease === "string"
      ? JSON.parse(lease) as unknown
      : Buffer.isBuffer(lease)
        ? JSON.parse(lease.toString("utf8")) as unknown
        : lease;
    const payload = jsonObject(envelope)?.payload;
    if (typeof payload !== "string" || payload.length === 0) return undefined;
    return jsonObject(JSON.parse(Buffer.from(payload, "base64url").toString("utf8")) as unknown);
  } catch (error) {
    return undefined;
  }
}

function deriveReleaseRoot(browserExecutable: string, artifactBrowserExecutable: unknown): string | undefined {
  if (typeof artifactBrowserExecutable !== "string" || !artifactBrowserExecutable.trim()) return undefined;
  const expectedParts = artifactBrowserExecutable.replaceAll("\\", "/").split("/").filter(Boolean);
  if (expectedParts.length === 0) return undefined;
  const actualPath = resolve(browserExecutable);
  const actualParts = actualPath.replaceAll("\\", "/").split("/").filter(Boolean);
  if (actualParts.length < expectedParts.length) return undefined;
  const actualSuffix = actualParts.slice(-expectedParts.length);
  const matches = expectedParts.every((part, index) => {
    const actualPart = actualSuffix[index];
    return typeof actualPart === "string" && part.toLowerCase() === actualPart.toLowerCase();
  });
  if (!matches) return undefined;
  let releaseRoot = actualPath;
  for (let index = 0; index < expectedParts.length; ++index) {
    releaseRoot = dirname(releaseRoot);
  }
  return releaseRoot;
}

function releaseRootFromLease(
  browserExecutable: string,
  lease: string | Buffer | Record<string, unknown>,
): string | undefined {
  const claims = parseSignedLeaseClaims(lease);
  const artifact = jsonObject(claims?.artifact);
  return deriveReleaseRoot(browserExecutable, artifact?.browserExecutable);
}

function validateFrameworkSettings(backend: FrameworkBackend, settings: FrameworkLaunchSettings): void {
  resolveFrameworkVersion(backend, settings.frameworkVersion);
  requireFrameworkHumanizeSupport(backend, settings.humanize);
  if (settings.nativeReady === true &&
      settings.nativeReadyTimeout !== undefined &&
      (!Number.isFinite(settings.nativeReadyTimeout) || settings.nativeReadyTimeout < 1)) {
    throw new ConfigurationError("nativeReadyTimeout must be a positive number of milliseconds", "config_invalid");
  }
}

function nativeHumanizeControl(
  backend: FrameworkBackend,
  settings: FrameworkLaunchSettings,
): Record<string, unknown> | undefined {
  if (settings.humanize !== true) return undefined;
  if (settings.humanSeed !== undefined &&
      (!Number.isInteger(settings.humanSeed) || settings.humanSeed < 0)) {
    throw new ConfigurationError("Native Humanize seed must be a non-negative integer", "humanize_seed_invalid");
  }
  const preset = settings.humanPreset ?? "default";
  return {
    schemaVersion: 1,
    kind: "slybrowser.native-humanize-control",
    backend,
    humanize: {
      enabled: true,
      version: 1,
      preset,
      config: resolveHumanConfig({
        preset,
        ...(settings.humanConfig === undefined ? {} : { config: settings.humanConfig }),
      }),
      ...(settings.humanSeed === undefined ? {} : { seed: settings.humanSeed }),
    },
  };
}

function mergeLaunchOptions(
  launchOptions: Record<string, unknown> | undefined,
  executable: string,
  handoffArguments: readonly string[],
): Record<string, unknown> {
  const result = { ...(launchOptions ?? {}) };
  if ("executablePath" in result || "executable_path" in result) {
    throw new ConfigurationError(
      "The browser executable must be passed to the SlyBrowser adapter",
      "executable_option_conflict",
    );
  }
  const existingArguments = result.args ?? [];
  if (!Array.isArray(existingArguments) || existingArguments.some((argument) => typeof argument !== "string")) {
    throw new ConfigurationError("Framework args must be a string array", "config_invalid");
  }
  return {
    ...result,
    executablePath: executable,
    args: [...handoffArguments, ...existingArguments],
  };
}

function nativeReadyTimeout(settings: FrameworkLaunchSettings): number | undefined {
  return settings.nativeReady === true ? settings.nativeReadyTimeout ?? 15_000 : undefined;
}

async function closeFrameworkRuntime(runtime: unknown): Promise<void> {
  const candidate = runtime as { close?: (...args: unknown[]) => unknown } | undefined;
  if (typeof candidate?.close !== "function") return;
  await Promise.resolve(candidate.close.call(runtime)).catch(() => undefined);
}

export async function launchPlaywright<TBrowser>(
  playwright: PlaywrightLike<TBrowser>,
  executable: string,
  lease: string | Buffer | Record<string, unknown>,
  settings: FrameworkLaunchSettings = {},
): Promise<TBrowser> {
  validateFrameworkSettings("playwright", settings);
  const readyTimeout = nativeReadyTimeout(settings);
  const plan = await prepareLaunch(executable, settings.profile ?? {}, lease, {
    tempRoot: settings.tempRoot,
    runtimeHandoff: settings.runtimeHandoff,
    allowRuntimeActivationTicket: settings.allowRuntimeActivationTicket === true,
    humanizeControl: nativeHumanizeControl("playwright", settings),
    nativeReady: readyTimeout !== undefined,
    releaseRoot: settings.releaseRoot === undefined ? releaseRootFromLease(executable, lease) : settings.releaseRoot,
  });
  try {
    const browser = await playwright.chromium.launch(
      mergeLaunchOptions(settings.launchOptions, plan.executable, plan.arguments),
    );
    try {
      if (readyTimeout !== undefined) await plan.waitForNativeReady(readyTimeout);
      return browser;
    } catch (error) {
      await closeFrameworkRuntime(browser);
      throw error;
    }
  } finally {
    await plan.cleanup();
  }
}

export async function launchPlaywrightPersistent<TContext>(
  playwright: PlaywrightLike<unknown, TContext>,
  userDataDir: string,
  executable: string,
  lease: string | Buffer | Record<string, unknown>,
  settings: FrameworkLaunchSettings = {},
): Promise<TContext> {
  validateFrameworkSettings("playwright", settings);
  const readyTimeout = nativeReadyTimeout(settings);
  const plan = await prepareLaunch(executable, settings.profile ?? {}, lease, {
    tempRoot: settings.tempRoot,
    runtimeHandoff: settings.runtimeHandoff,
    allowRuntimeActivationTicket: settings.allowRuntimeActivationTicket === true,
    humanizeControl: nativeHumanizeControl("playwright", settings),
    nativeReady: readyTimeout !== undefined,
    releaseRoot: settings.releaseRoot === undefined ? releaseRootFromLease(executable, lease) : settings.releaseRoot,
  });
  try {
    const context = await playwright.chromium.launchPersistentContext(
      resolve(userDataDir),
      mergeLaunchOptions(settings.launchOptions, plan.executable, plan.arguments),
    );
    try {
      if (readyTimeout !== undefined) await plan.waitForNativeReady(readyTimeout);
      return context;
    } catch (error) {
      await closeFrameworkRuntime(context);
      throw error;
    }
  } finally {
    await plan.cleanup();
  }
}

export async function launchPuppeteer<TBrowser>(
  puppeteer: PuppeteerLike<TBrowser>,
  executable: string,
  lease: string | Buffer | Record<string, unknown>,
  settings: FrameworkLaunchSettings = {},
): Promise<TBrowser> {
  validateFrameworkSettings("puppeteer", settings);
  const readyTimeout = nativeReadyTimeout(settings);
  const plan = await prepareLaunch(executable, settings.profile ?? {}, lease, {
    tempRoot: settings.tempRoot,
    runtimeHandoff: settings.runtimeHandoff,
    allowRuntimeActivationTicket: settings.allowRuntimeActivationTicket === true,
    humanizeControl: nativeHumanizeControl("puppeteer", settings),
    nativeReady: readyTimeout !== undefined,
    releaseRoot: settings.releaseRoot === undefined ? releaseRootFromLease(executable, lease) : settings.releaseRoot,
  });
  try {
    const browser = await puppeteer.launch(
      mergeLaunchOptions(settings.launchOptions, plan.executable, plan.arguments),
    );
    try {
      if (readyTimeout !== undefined) await plan.waitForNativeReady(readyTimeout);
      return browser;
    } catch (error) {
      await closeFrameworkRuntime(browser);
      throw error;
    }
  } finally {
    await plan.cleanup();
  }
}

export async function launchPuppeteerPersistent<TBrowser>(
  puppeteer: PuppeteerLike<TBrowser>,
  userDataDir: string,
  executable: string,
  lease: string | Buffer | Record<string, unknown>,
  settings: FrameworkLaunchSettings = {},
): Promise<TBrowser> {
  if (settings.launchOptions && "userDataDir" in settings.launchOptions) {
    throw new ConfigurationError(
      "The persistent profile directory must be passed to the SlyBrowser adapter",
      "profile_option_conflict",
    );
  }
  return launchPuppeteer(puppeteer, executable, lease, {
    ...settings,
    launchOptions: {
      ...(settings.launchOptions ?? {}),
      userDataDir: resolve(userDataDir),
    },
  });
}

File: packages/python/src/slybrowser/browser.py
"""Framework adapters that launch the native SlyBrowser executable."""

from __future__ import annotations

from pathlib import Path
from typing import Any, Mapping

from .automation import require_playwright_humanize_support, resolve_playwright_version
from .errors import ConfigurationError
from .launcher import LaunchPlan, prepare_launch, wait_for_native_ready
from .webdriver import _release_root_from_lease, resolve_humanize_config


def _launch_arguments(
    launch_options: Mapping[str, Any] | None,
    handoff_arguments: tuple[str, ...],
) -> dict[str, Any]:
    result = dict(launch_options or {})
    if "executable_path" in result or "executablePath" in result:
        raise ConfigurationError(
            "The browser executable must be passed to the SlyBrowser adapter",
            code="executable_option_conflict",
        )
    existing_arguments = result.get("args", [])
    if not isinstance(existing_arguments, (list, tuple)) or any(
        not isinstance(argument, str) for argument in existing_arguments
    ):
        raise ConfigurationError("Playwright args must be a string array", code="config_invalid")
    result["args"] = [*handoff_arguments, *existing_arguments]
    return result


def _camel_humanize_config(config: object) -> dict[str, int]:
    return {
        "mouseStepsMin": config.mouse_steps_min,
        "mouseStepsMax": config.mouse_steps_max,
        "mouseStepDelayMin": config.mouse_step_delay_min,
        "mouseStepDelayMax": config.mouse_step_delay_max,
        "clickHoldMin": config.click_hold_min,
        "clickHoldMax": config.click_hold_max,
        "keyDelayMin": config.key_delay_min,
        "keyDelayMax": config.key_delay_max,
        "thinkDelayMin": config.think_delay_min,
        "thinkDelayMax": config.think_delay_max,
    }


def _native_humanize_control(
    *,
    humanize: bool,
    human_preset: str,
    human_config: Mapping[str, int | float] | None,
    human_seed: int | None,
) -> dict[str, Any] | None:
    if not humanize:
        return None
    if human_seed is not None and (not isinstance(human_seed, int) or human_seed < 0):
        raise ConfigurationError("Native Humanize seed must be a non-negative integer", code="humanize_seed_invalid")
    return {
        "schemaVersion": 1,
        "kind": "slybrowser.native-humanize-control",
        "backend": "playwright",
        "humanize": {
            "enabled": True,
            "version": 1,
            "preset": human_preset,
            "config": _camel_humanize_config(resolve_humanize_config(human_preset, human_config)),
            **({} if human_seed is None else {"seed": human_seed}),
        },
    }


def _validate_native_ready(native_ready: bool, native_ready_timeout: float) -> None:
    if native_ready and (not isinstance(native_ready_timeout, (int, float)) or native_ready_timeout <= 0):
        raise ConfigurationError("native_ready_timeout must be positive", code="config_invalid")


def _close_runtime(runtime: Any) -> None:
    close = getattr(runtime, "close", None)
    if callable(close):
        try:
            close()
        except Exception:
            pass


async def _close_runtime_async(runtime: Any) -> None:
    import inspect

    close = getattr(runtime, "close", None)
    if callable(close):
        try:
            result = close()
            if inspect.isawaitable(result):
                await result
        except Exception:
            pass


def _wait_if_requested(plan: LaunchPlan, native_ready: bool, native_ready_timeout: float) -> None:
    if native_ready:
        wait_for_native_ready(plan, timeout=native_ready_timeout)


def launch_playwright(
    playwright: Any,
    executable: str | Path,
    lease: str | bytes | Mapping[str, Any],
    *,
    profile: Mapping[str, Any] | None = None,
    launch_options: Mapping[str, Any] | None = None,
    temp_root: str | Path | None = None,
    framework_version: str | None = None,
    humanize: bool = False,
    human_preset: str = "default",
    human_config: Mapping[str, int | float] | None = None,
    human_seed: int | None = None,
    runtime_handoff: Mapping[str, Any] | None = None,
    allow_runtime_activation_ticket: bool = False,
    release_root: str | Path | None = None,
    native_ready: bool = False,
    native_ready_timeout: float = 15.0,
) -> Any:
    """Launch a non-persistent Playwright browser using an existing Playwright object."""

    resolve_playwright_version(framework_version)
    require_playwright_humanize_support(humanize)
    _validate_native_ready(native_ready, native_ready_timeout)
    with prepare_launch(
        executable,
        profile or {},
        lease,
        temp_root=temp_root,
        runtime_handoff=runtime_handoff,
        allow_runtime_activation_ticket=allow_runtime_activation_ticket,
        release_root=Path(release_root).expanduser().resolve() if release_root is not None else _release_root_from_lease(executable, lease),
        native_ready=native_ready,
        humanize_control=_native_humanize_control(
            humanize=humanize,
            human_preset=human_preset,
            human_config=human_config,
            human_seed=human_seed,
        ),
    ) as plan:
        options = _launch_arguments(launch_options, plan.arguments)
        runtime = playwright.chromium.launch(executable_path=str(plan.executable), **options)
        try:
            _wait_if_requested(plan, native_ready, native_ready_timeout)
            return runtime
        except BaseException:
            _close_runtime(runtime)
            raise


async def launch_playwright_async(
    playwright: Any,
    executable: str | Path,
    lease: str | bytes | Mapping[str, Any],
    *,
    profile: Mapping[str, Any] | None = None,
    launch_options: Mapping[str, Any] | None = None,
    temp_root: str | Path | None = None,
    framework_version: str | None = None,
    humanize: bool = False,
    human_preset: str = "default",
    human_config: Mapping[str, int | float] | None = None,
    human_seed: int | None = None,
    runtime_handoff: Mapping[str, Any] | None = None,
    allow_runtime_activation_ticket: bool = False,
    release_root: str | Path | None = None,
    native_ready: bool = False,
    native_ready_timeout: float = 15.0,
) -> Any:
    resolve_playwright_version(framework_version)
    require_playwright_humanize_support(humanize)
    _validate_native_ready(native_ready, native_ready_timeout)
    with prepare_launch(
        executable,
        profile or {},
        lease,
        temp_root=temp_root,
        runtime_handoff=runtime_handoff,
        allow_runtime_activation_ticket=allow_runtime_activation_ticket,
        release_root=Path(release_root).expanduser().resolve() if release_root is not None else _release_root_from_lease(executable, lease),
        native_ready=native_ready,
        humanize_control=_native_humanize_control(
            humanize=humanize,
            human_preset=human_preset,
            human_config=human_config,
            human_seed=human_seed,
        ),
    ) as plan:
        options = _launch_arguments(launch_options, plan.arguments)
        runtime = await playwright.chromium.launch(executable_path=str(plan.executable), **options)
        try:
            _wait_if_requested(plan, native_ready, native_ready_timeout)
            return runtime
        except BaseException:
            await _close_runtime_async(runtime)
            raise


def launch_playwright_persistent(
    playwright: Any,
    user_data_dir: str | Path,
    executable: str | Path,
    lease: str | bytes | Mapping[str, Any],
    *,
    profile: Mapping[str, Any] | None = None,
    launch_options: Mapping[str, Any] | None = None,
    temp_root: str | Path | None = None,
    framework_version: str | None = None,
    humanize: bool = False,
    human_preset: str = "default",
    human_config: Mapping[str, int | float] | None = None,
    human_seed: int | None = None,
    runtime_handoff: Mapping[str, Any] | None = None,
    allow_runtime_activation_ticket: bool = False,
    release_root: str | Path | None = None,
    native_ready: bool = False,
    native_ready_timeout: float = 15.0,
) -> Any:
    resolve_playwright_version(framework_version)
    require_playwright_humanize_support(humanize)
    _validate_native_ready(native_ready, native_ready_timeout)
    with prepare_launch(
        executable,
        profile or {},
        lease,
        temp_root=temp_root,
        runtime_handoff=runtime_handoff,
        allow_runtime_activation_ticket=allow_runtime_activation_ticket,
        release_root=Path(release_root).expanduser().resolve() if release_root is not None else _release_root_from_lease(executable, lease),
        native_ready=native_ready,
        humanize_control=_native_humanize_control(
            humanize=humanize,
            human_preset=human_preset,
            human_config=human_config,
            human_seed=human_seed,
        ),
    ) as plan:
        options = _launch_arguments(launch_options, plan.arguments)
        runtime = playwright.chromium.launch_persistent_context(
            str(Path(user_data_dir).expanduser().resolve()),
            executable_path=str(plan.executable),
            **options,
        )
        try:
            _wait_if_requested(plan, native_ready, native_ready_timeout)
            return runtime
        except BaseException:
            _close_runtime(runtime)
            raise


async def launch_playwright_persistent_async(
    playwright: Any,
    user_data_dir: str | Path,
    executable: str | Path,
    lease: str | bytes | Mapping[str, Any],
    *,
    profile: Mapping[str, Any] | None = None,
    launch_options: Mapping[str, Any] | None = None,
    temp_root: str | Path | None = None,
    framework_version: str | None = None,
    humanize: bool = False,
    human_preset: str = "default",
    human_config: Mapping[str, int | float] | None = None,
    human_seed: int | None = None,
    runtime_handoff: Mapping[str, Any] | None = None,
    allow_runtime_activation_ticket: bool = False,
    release_root: str | Path | None = None,
    native_ready: bool = False,
    native_ready_timeout: float = 15.0,
) -> Any:
    resolve_playwright_version(framework_version)
    require_playwright_humanize_support(humanize)
    _validate_native_ready(native_ready, native_ready_timeout)
    with prepare_launch(
        executable,
        profile or {},
        lease,
        temp_root=temp_root,
        runtime_handoff=runtime_handoff,
        allow_runtime_activation_ticket=allow_runtime_activation_ticket,
        release_root=Path(release_root).expanduser().resolve() if release_root is not None else _release_root_from_lease(executable, lease),
        native_ready=native_ready,
        humanize_control=_native_humanize_control(
            humanize=humanize,
            human_preset=human_preset,
            human_config=human_config,
            human_seed=human_seed,
        ),
    ) as plan:
        options = _launch_arguments(launch_options, plan.arguments)
        runtime = await playwright.chromium.launch_persistent_context(
            str(Path(user_data_dir).expanduser().resolve()),
            executable_path=str(plan.executable),
            **options,
        )
        try:
            _wait_if_requested(plan, native_ready, native_ready_timeout)
            return runtime
        except BaseException:
            await _close_runtime_async(runtime)
            raise
